core/services/relay/evm/evm_service.go

```diff
@@ -21,9 +22,38 @@ import (
 	"github.com/smartcontractkit/chainlink-framework/chains"
 )
 
+// resolveBlock converts readAt into an explicit block number.
+// readAt may be either:
+//   - a decimal-encoded big int block number (e.g. "17")
+//   - a confidence level enum
+func (r *Relayer) resolveBlock(ctx context.Context, readAt string) (*big.Int, error) {
+	if blockNum, ok := new(big.Int).SetString(readAt, 10); ok {
+		return blockNum, nil
+	}
+
+	latest, finalized, err := r.LatestAndFinalizedHead(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch block number for readAt: %q, err: %w", readAt, err)
+	}
+
+	switch primitives.ConfidenceLevel(readAt) {
+	case primitives.Finalized:
+		return finalized.Number, nil
+	case primitives.Unconfirmed:
+		return latest.Number, nil
+	default:
+		return nil, fmt.Errorf("unsupported confidence level for readAt: %q", readAt)
+	}
+}
+
 // Direct RPC
-func (r *Relayer) CallContract(ctx context.Context, msg *evmtypes.CallMsg, blockNumber *big.Int) ([]byte, error) {
-	return r.chain.Client().CallContract(ctx, toEthMsg(msg), blockNumber)
+func (r *Relayer) CallContract(ctx context.Context, msg *evmtypes.CallMsg, readAt string) ([]byte, error) {
```

Comment:
Shouldn't we try to be as close to the RPC client as possible? 
It solves the same problem by accepting `*big.Int`. If it's nil, it converts it to "latest"; if it's `-3` - to "finalized". Should not we use the same approach to make it less confusing for a user who is more likely to be familiar with the RPC Client SDK than with our internal definitions?

Reply:
I thought the point was to implement this ^ interface with a thin adapter in the sdk, no?